clarity/contracts/arkadiko-oracle-v1-1.clar

```diff
@@ -18,28 +30,225 @@
   }
 )
 
-(define-public (set-oracle-owner (address principal))
-  (begin
-    (asserts! (is-eq tx-sender (var-get oracle-owner)) (err ERR-NOT-AUTHORIZED))
+(define-map trusted-oracles (buff 33) bool)
 
-    (ok (var-set oracle-owner address))
-  )
+(define-map token-id-to-names uint (list 4 (string-ascii 12)))
+(define-map token-name-to-id (string-ascii 12) uint)
+
+;; ---------------------------------------------------------
+;; Getters
+;; ---------------------------------------------------------
+
+;; @desc check if given public key is trusted
+(define-read-only (is-trusted-oracle (pubkey (buff 33)))
+  (default-to false (map-get? trusted-oracles pubkey))
 )
 
-(define-public (update-price (token (string-ascii 12)) (price uint) (decimals uint))
-  (if (is-eq tx-sender (var-get oracle-owner))
-    (begin
-      (map-set prices { token: token } { last-price: price, last-block: block-height, decimals: decimals })
-      (ok price)
-    )
-    (err ERR-NOT-WHITELISTED)
-  )
+;; @desc get token ID for given name
+(define-read-only (get-token-id-from-name (name (string-ascii 12)))
+  (default-to u0 (map-get? token-name-to-id name))
 )
 
+;; @desc get list of token names for given token ID
+(define-read-only (get-token-names-from-id (id uint))
+  (default-to (list ) (map-get? token-id-to-names id))
+)
+
+;; @desc get minimum valid signers needed
+(define-read-only (get-minimum-valid-signers)
+  (var-get minimum-valid-signers)
+)
+
+;; @desc get price info for given token name
 (define-read-only (get-price (token (string-ascii 12)))
   (unwrap! (map-get? prices { token: token }) { last-price: u0, last-block: u0, decimals: u0 })
 )
 
+;; @desc get price info response for given token name
 (define-public (fetch-price (token (string-ascii 12)))
   (ok (get-price token))
 )
+
+;; ---------------------------------------------------------
+;; Message signing
+;; ---------------------------------------------------------
+
+;; @desc update token price as DAO owner
+;; @param token; token to update price for
+;; @param price; price value
+;; @param decimals; amount of decimals
+;; @post bool; returns true
+(define-public (update-price-owner (token (string-ascii 12)) (price uint) (decimals uint))
+  (begin
+    (asserts! (is-eq tx-sender (contract-call? .arkadiko-dao get-dao-owner)) (err ERR-NOT-AUTHORIZED))
+    (map-set prices { token: token } { last-price: price, last-block: block-height, decimals: decimals })
+    (ok true)
+  )
+)
+
+;; @desc update token price as multisig node
+;; @param block; the block on which the message was created
+;; @param token-id; token ID to update price for
+;; @param price; price value
+;; @param decimals; amount of decimals
+;; @param signatures; list of signatures from oracle nodes
+;; @post bool; returns true if successful
+(define-public (update-price-multi (block uint) (token-id uint) (price uint) (decimals uint) (signatures (list 10 (buff 65))))
```

Comment:
Q: can you explain how this is being called by 3 separate nodes and only then updated? I don't get this? It seems like this method only needs to be called once with the right signatures? How is it possible to spread this among servers then?

Reply:
A: Correct, it needs to be called once with at least 3 correct signatures to update the price. 

The idea is that we will run 5 different nodes and they each check if the price needs to be updated based on the last  update block (and if other nodes have not pushed an update in the mempool). If an update is needed the node will fetch the price from it's source (we have 5 different sources for STX and BTC, the other tokens are fetched directly from the AMMs on chain). The node will ask the other nodes to sign the price data (block, token ID, price, decimals). The other nodes will only do this if the provided input is correct (price is within certain range of the price of their source). If the node managed to get 3 or more signatures, it pushes the transaction on chain.

I have shared the oracle node repo as well.

Reply:
@nieldeckx I see, thanks, so a lot of logic is ran off-chain.

Can you share the link of the Oracle node repo here? Can't see or find it

Reply:
@philipdesmedt https://github.com/arkadiko-dao/oracle-node (private repo for now)